fixed_bed_model.py

```diff
@@ -0,0 +1,2630 @@
+    '''
+    li = np.asarray(lis)
+    res = li.reshape([69*20,])
+    return sum(res)/len(res)
+
+def time_points(n, time_range):
+    '''
+    Split the timespan of the experiment into n nodes
+    
+    Argument: n: the number of time nodes.
+    
+    return: time_points, a list of values of timepoints in [s]
+    '''
+    time_points = []
+    int_time = time_range/n
+    
+    # specify where the time begins. Usually 0
+    time_ele = 0
+    time_points.append(time_ele)
+    for i in range(n):
+        time_ele += int_time
+        time_points.append(time_ele)
+
+    return time_points
+
+def Residual(m):
+    '''
+    Calculate the residual with a given model
+    return: 
+        res: the residual at every timepoint
+        FCO2: the CO2 outlet flowrate at every timepoint
+    '''
+    
+    # Extract time 
+    T = []
+    for j in m.t:
+        T.append(value(j))
+        t_final = T[-1]
+        
+    # Extract CO2 density 
+    outlet_den = extract3d(m, m.C, 'CO2')
+    outlet_ = []
+    outlet_.append(outlet_den[Ngrid-1, :])
+    outlet = np.reshape(outlet_, len(T))
+    
+    # Extract velocity 
+    outlet_vel = extract2d(m, m.v)
+    outlet_v = []
+    outlet_v.append(outlet_vel[Ngrid-1, :])
+    outlet_velo = np.reshape(outlet_v, len(T))
+
+    # Extract values of FCO2
+    FCO2_ = []
+    for i in range(0, len(T)):
+        FCO2_.append(value(m.FCO2[T[i]]))
+    
+    # Extract values of yCO2
+    yCO2_ = []
+    for i in range(0, len(T)):
+        yCO2_.append(value(m.yCO2[T[i]]))
+     
+    # Get value of residual
+    res = 0
+    for i in range(0, len(outlet)):
+        res += (outlet[i]*outlet_velo[i]*3.1415926*rbed*rbed*600 - FCO2_[i])**2
+        
+    FCO2 = np.zeros((len(outlet)))
+    for i in range(0, len(outlet)):
+        FCO2[i] = (outlet[i]*outlet_velo[i]*3.1415926*rbed*rbed*600)
+    
+    return res, FCO2
+
+    
+def breakthrough_modify2(m, file=None, source="computer"):
+    '''
+    Draw breakthrough curve of the Pyomo model
+    
+    Arguments:
+        m: Pyomo model
+        file: when source = 'computer', this is where the computer experimental data is stored. Otherwise, its default is None
+        source: if computer, it is plotting the computer experiment data. If lab, its comparing the experiments from the lab
+    
+    Return: None 
+    
+    Other: plot
+    
+    '''
+    
+    
+    # unit: mol/m3
+    outlet_den, _, _, _ = extract3d(m, m.C, 'CO2')
+    
+    # unit: cm/s
+    outlet_vel, _, _, _ = extract2d(m, m.v)
+    
+    outlet_n2, _, _, _ = extract3d(m, m.C, 'N2')
+    #print(outlet_den)
+    #print(outlet_n2)
+    #print(outlet_vel)
+    
+    model_temp, _,_,_ = extract2d(m,m.temp)
+    
+    
+    T = []
+    for j in m.t:
+        T.append(value(j))
+        t_final = T[-1]
+        
+    #print(T)
+    #print(exp['time']*60)
+    #print(break_wvu['time']*60)
+    
+    if source == "lab":
+        break_wvu = pd.read_csv('breakthrough_wvu.csv')
```

Comment:
Remove hardcoded references to file names. Can this code be generalized? Is it needed?

Reply:
This 'breakthrough_wvu' file contains the experimental data from WVU, which I draw together with our simulation curve for easy comparison. I think just removing it is fine since we do not need to show this experimental data anyway.